akka-stream/src/main/scala/akka/stream/impl/fusing/MapAsyncPartitioned.scala

```diff
@@ -127,7 +127,13 @@ private[akka] final case class MapAsyncPartitioned[In, Out, Partition](
                 // dropped from per-partition queue, to be able to execute next, but result is kept in the
                 // main linear buffer for when out is ready
                 buffer.dropOnlyPartitionHead(partition)
-                dropCompletedThenPushIfPossible(partition)
+
+                if (perPartition > 1 && buffer.usedInPartition(partition) > perPartition) {
```

Comment:
The special case for `perPartition == 1` does mean that we assume that any future started as a consequence of the `dropOnlyPartitionHead` above will wait until that async callback triggers.

Not only do we save the cost of introspecting the partition buffer's `used`, we also save the cost of an almost certainly pointless iteration of peeking and checking for completion.